### DISCUSSION TITLE
Is it possible to get the generated bindings as js_sys::Functions?

### DISCUSSION THREAD

**Discussion by allsey87** [wasm-bindgen/wasm-bindgen#3491] [Q&A]
I am working on some WebAssembly in WebAssembly type code, where I use web-sys and js-sys to instansiate other WebAssembly modules. For this reason, I would find it really useful if I were able to pass the functions that I have written in Rust to other WebAssembly modules. The following code is a trivial example of what I would like to achieve:

```rust
#[wasm_bindgen]
pub fn memory_saturating_add(memory: js_sys::WebAssembly::Memory, index: u32, value: u8) {
    let buffer = js_sys::Uint8Array::new(&memory.buffer());
    buffer.set_index(index, buffer.get_index(index).saturating_add(value));
}
```

```rust
fn create_wasm_imports(&self) -> js_sys::Object {
    let imports = js_sys::Object::new();

    // get the wasm-bindgen js function using hypothetical bindings function
    let function: js_sys::Function = wasm_bindgen::bindings()
        .get("memory_saturating_add")
        .unwrap();

    // bind the memory so that we have a function with only native wasm
    // arguments which can be called from a WebAssembly module
    let function = function.bind(self.memory);
    
    js_sys::Reflect::set(&imports, &"_saturating_add".into(), &function).unwrap();
    imports
}
```

Note that:
1. `wasm_bindgen::exports` gives us access to a `memory_saturating_add` but it is the raw WebAssembly export and cannot be used if we need to bind that first argument using a `JsValue`.
2. I also tried to reimport the exported js function as follows but it seems to cause wasm-bindgen to rename the functions, even if I use `js_name` attribute:

```rust
#[wasm_bindgen]
extern "C" {
    fn memory_saturating_add(memory: js_sys::WebAssembly::Memory, index: u32, value: u8);
}
```

Any thoughts on my approaches here? Does this make sense?

**Comment by daxpedda** [comment]
Very interesting use-case!

Unfortunately this is definitely not currently possible.
I think it shouldn't be too hard to expose something like `wasm_bindgen::js_exports()`.

I'm happy to review a PR!

**Reply by allsey87** [reply]
If you have some ideas/pointers on how this could be implemented, I can take a shot at it

**Reply by daxpedda** [reply]
You basically have to copy what has been done here: https://github.com/rustwasm/wasm-bindgen/pull/3296.
Don't remember exactly what you have to expose on the JS side, but if you look at the generated JS shim it shouldn't be too hard to figure out.
